// typia/lib/utils/RandomGenerator.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RandomGenerator = void 0;
var randexp_1 = __importDefault(require("randexp"));
var RandomGenerator;
(function (RandomGenerator) {
    var ALPHABETS = "abcdefghijklmnopqrstuvwxyz";
    function boolean() {
        return Math.random() < 0.5;
    }
    RandomGenerator.boolean = boolean;
    function integer(min, max) {
        min !== null && min !== void 0 ? min : (min = 0);
        max !== null && max !== void 0 ? max : (max = 100);
        return Math.floor(Math.random() * (max - min + 1)) + min;
    }
    RandomGenerator.integer = integer;
    function bigint(min, max) {
        min !== null && min !== void 0 ? min : (min = BigInt(0));
        max !== null && max !== void 0 ? max : (max = BigInt(100));
        return BigInt(integer(Number(min), Number(max)));
    }
    RandomGenerator.bigint = bigint;
    function number(min, max) {
        min !== null && min !== void 0 ? min : (min = 0);
        max !== null && max !== void 0 ? max : (max = 100);
        return Math.random() * (max - min) + min;
    }
    RandomGenerator.number = number;
    function string(length) {
        return new Array(length !== null && length !== void 0 ? length : integer(5, 10))
            .fill(0)
            .map(function () { return ALPHABETS[integer(0, ALPHABETS.length - 1)]; })
            .join("");
    }
    RandomGenerator.string = string;
    function array(closure, count) {
        return new Array(count !== null && count !== void 0 ? count : integer(0, 3))
            .fill(0)
            .map(function (_e, index) { return closure(index); });
    }
    RandomGenerator.array = array;
    function pick(array) {
        return array[integer(0, array.length - 1)];
    }
    RandomGenerator.pick = pick;
    function uuid() {
        return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, function (c) {
            var r = (Math.random() * 16) | 0;
            var v = c === "x" ? r : (r & 0x3) | 0x8;
            return v.toString(16);
        });
    }
    RandomGenerator.uuid = uuid;
    function email() {
        return "".concat(string(10), "@").concat(string(10), ".").concat(string(3));
    }
    RandomGenerator.email = email;
    function url() {
        return "https://".concat(string(10), ".").concat(string(3));
    }
    RandomGenerator.url = url;
    function ipv4() {
        return array(function () { return integer(0, 255); }, 4).join(".");
    }
    RandomGenerator.ipv4 = ipv4;
    function ipv6() {
        return array(function () { return integer(0, 65535).toString(16); }, 8).join(":");
    }
    RandomGenerator.ipv6 = ipv6;
    function pattern(regex) {
        return new randexp_1.default(regex).gen();
    }
    RandomGenerator.pattern = pattern;
    function date(min, max) {
        min !== null && min !== void 0 ? min : (min = 0);
        max !== null && max !== void 0 ? max : (max = Date.now() * 2);
        return new Date(number(min, max)).toISOString().substring(0, 10);
    }
    RandomGenerator.date = date;
    function datetime(min, max) {
        min !== null && min !== void 0 ? min : (min = 0);
        max !== null && max !== void 0 ? max : (max = Date.now() * 2);
        return new Date(number(min, max)).toISOString();
    }
    RandomGenerator.datetime = datetime;
})(RandomGenerator = exports.RandomGenerator || (exports.RandomGenerator = {}));
