// typia/lib/executable/setup/ArgumentParser.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ArgumentParser = void 0;
var path_1 = __importDefault(require("path"));
var FileRetriever_1 = require("./FileRetriever");
var ArgumentParser;
(function (ArgumentParser) {
    var _this = this;
    ArgumentParser.parse = function (pack) {
        return function (erase) {
            return function (inquiry) { return __awaiter(_this, void 0, void 0, function () {
                var newbie, command, prompt, action, output;
                var _this = this;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            newbie = {
                                commander: pack.install({
                                    dev: true,
                                    modulo: "commander",
                                    version: "10.0.0",
                                    silent: true,
                                }),
                                inquirer: pack.install({
                                    dev: true,
                                    modulo: "inquirer",
                                    version: "8.2.5",
                                    silent: true,
                                }),
                            };
                            return [4, FileRetriever_1.FileRetriever.require(path_1.default.join("node_modules", "commander"))(pack.directory)];
                        case 1:
                            command = (_a.sent()).program;
                            return [4, FileRetriever_1.FileRetriever.require(path_1.default.join("node_modules", "inquirer"))(pack.directory)];
                        case 2:
                            prompt = (_a.sent()).createPromptModule;
                            action = function (closure) {
                                return new Promise(function (resolve, reject) {
                                    command.action(function (options) { return __awaiter(_this, void 0, void 0, function () {
                                        var _a, exp_1;
                                        return __generator(this, function (_b) {
                                            switch (_b.label) {
                                                case 0:
                                                    _b.trys.push([0, 2, , 3]);
                                                    _a = resolve;
                                                    return [4, closure(options)];
                                                case 1:
                                                    _a.apply(void 0, [_b.sent()]);
                                                    return [3, 3];
                                                case 2:
                                                    exp_1 = _b.sent();
                                                    reject(exp_1);
                                                    return [3, 3];
                                                case 3: return [2];
                                            }
                                        });
                                    }); });
                                    command.parseAsync().catch(reject);
                                });
                            };
                            return [4, (function () { return __awaiter(_this, void 0, void 0, function () {
                                    var error_1;
                                    return __generator(this, function (_a) {
                                        switch (_a.label) {
                                            case 0:
                                                _a.trys.push([0, 2, , 3]);
                                                return [4, inquiry(pack, command, prompt, action)];
                                            case 1: return [2, _a.sent()];
                                            case 2:
                                                error_1 = _a.sent();
                                                return [2, error_1];
                                            case 3: return [2];
                                        }
                                    });
                                }); })()];
                        case 3:
                            output = _a.sent();
                            if (erase === true) {
                                if (newbie.commander)
                                    pack.erase({ modulo: "commander", silent: true });
                                if (newbie.inquirer)
                                    pack.erase({ modulo: "inquirer", silent: true });
                            }
                            if (output instanceof Error)
                                throw output;
                            return [2, output];
                    }
                });
            }); };
        };
    };
})(ArgumentParser = exports.ArgumentParser || (exports.ArgumentParser = {}));
